### PR TITLE
Date based releases

### PR DESCRIPTION
**Pull request by zklaus**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

RapidJSON has had its last release in 2016. However, development has never really stopped with hundreds of commits since then, seemingly focused on bugfixing and future proofing the library, e.g. by supporting newer architectures and c++ standards.

With this PR I propose that we follow the lead of other packagers like vcpkg [1,2] (shoutout to @timkpaine for providing the links), packaging date based releases from upstream `main`. The mechanism proposed in the updated meta.yaml is a calver based versioning with YYYY.MM.DD versions and a manually synced corresponding commit hash.

What do you think, @conda-forge/rapidjson?

[1] https://github.com/microsoft/vcpkg/blob/master/ports/rapidjson/vcpkg.json#L3
[2] https://github.com/microsoft/vcpkg/blob/master/ports/rapidjson/portfile.cmake#L5

### CODE REVIEW

**Comment by zklaus** [comment]
@conda-forge-admin, please rerender

File: recipe/meta.yaml [hunk 1]
```diff
@@ -1,18 +1,18 @@
...
 
 package:
   name: {{ name|lower }}
   version: {{ version }}
 
 source:
-  fn: {{ name }}-{{ version }}.tar.gz
-  url: https://github.com/miloyip/rapidjson/archive/v{{ version }}.tar.gz
-  sha256: {{ sha256 }}
+  git_url: https://github.com/Tencent/rapidjson.git
```
Comment (xhochy) [inline recipe/meta.yaml]: Let's please use GitHub tarballs and not plain git clones for this.

**Review by xhochy — CHANGES_REQUESTED** [review]
I would be OK with this if we upload the new releases to a label, e.g. `conda-forge/label/rapidjson_dev`

File: recipe/meta.yaml (hunk 1 above)
Comment (zklaus) [reply recipe/meta.yaml]: I thought that was only available for tags, but of course it works for commits too. Fixed now.

**Comment by zklaus** [comment]
> I would be OK with this if we upload the new releases to a label, e.g. `conda-forge/label/rapidjson_dev`

I understand your point and would be completely on-board in the usual circumstances where this situation would signify a development version with new features that is less tested than the stable release.
However, rapidjson is really different here. Basically, the maintainer has decided that releases are not for him, despite numerous pleas from the community. However, otherwise the development seems to be done rather conservatively with most commits since v1.1.0 concerned with bugfixes and maintenance, such as updates to ensure compiler compatibility.

I attach a poor-mans changelog composed of a oneline git log of commits since v1.1.0:
[rapidjson-changes-since-1.1.0.txt](https://github.com/conda-forge/rapidjson-feedstock/files/15016168/rapidjson-changes-since-1.1.0.txt).

That's why other package managers such as vcpkg have adopted date based packages into their mainline channels and why I would advocate for that here as well.

**Comment by xhochy** [comment]
My main concern with putting them on the main label is that once you went to date based labels, you cannot go back anymore.

**Comment by timkpaine** [comment]
@xhochy agreed, but `vcpkg`, `conan`, `homebrew`, and more have decided that newer release is better. If and when a new version is released, that bridge can be crossed (perhaps by yanking previous date based releases).

**Comment by zklaus** [comment]
> My main concern with putting them on the main label is that once you went to date based labels, you cannot go back anymore.

That's a valid concern. Perhaps there are ways around? I guess the canonical way in some sense would be [epochs](https://packaging.python.org/en/latest/specifications/version-specifiers/#version-epochs) (though I am not sure about the state of support for this in conda-forge). Another possibility would be to see this as a continuation of the current 1.1.0 release, perhaps simply with versions `1.1.0.YYYYMMDD`?  Any official new version will likely be at least `1.2.0`.

@xhochy, do you think something along these lines could work?

**Comment by timkpaine** [comment]
@zklaus I do like the idea of `1.1.0.YYYYMMDD`, this seems to solve both problems (allow new versions but protect against future releases). Hopefully this version scheme wouldn't cause any solver issues or anything?

**Comment by xhochy** [comment]
> @xhochy agreed, but vcpkg, conan, homebrew, and more have decided that newer release is better.

I don't see the new release in `homebrew`. This would have been the most authoritative source of a new versioning scheme for me.

**Comment by zklaus** [comment]
Looks like homebrew indeed offers 1.1.0 as stable, though they also provide `HEAD`. Microsoft's vcpkg does date based releases.

But apart from what others do, it's illustrative to look at the upstream issues on this, for example the relatively recent https://github.com/Tencent/rapidjson/issues/2238, which in a comment lists no less than 18 other issues in which the lack of an official release was discussed, and not only as an academic problem, but actual real bugs hurting people, cf https://github.com/Tencent/rapidjson/issues/1006#issuecomment-1416688071.

So I think it's highly reasonable to provide a newer version, and to be honest I don't feel much disagreement on that point.

How can we do so without blocking easy adoption of a possible new release and with an optimal user experience?
I think we can adopt a date based versioning scheme, e.g. 1.1.0.YYYYMMDD, or perhaps 1.1.0.postYYYYMMDD (shoutout to @jaimergp, with whom I discussed the issue and who proposed the post version as a candidate). Both would be allowed and serve our needs according to the conda-build docs, which say the version should follow [PEP-386, verlib](https://peps.python.org/pep-0386/#the-new-versioning-algorithm).

Additionally, we could put these new releases on a label like `rapidjson_dev`. That would keep the stable release in main as the default. However, as discussed, imho there is reason to believe that the "stable" release is not actually any more stable than a date based more recent version; quite the opposite.

@xhochy, could you help me understand the user experience implications of putting things on a dev label? Would it be possible for recipe authors to opt-in to the newer releases? Or would that always impose end-user action to activate the dev label?

Finally, another option (again, thanks @jaimergp!) would be to make a wholly new package, say `rapidjson-dev` or `rapidjson-post` or some such which from the get go would just use date based releases. That would give a straightforward opt-in for recipe authors.

**Comment by xhochy** [comment]
> @xhochy, could you help me understand the user experience implications of putting things on a dev label? Would it be possible for recipe authors to opt-in to the newer releases? Or would that always impose end-user action to activate the dev label?

The respective feedstock would need the following bit in their `recipe/conda_build_config.yaml`:

```
channel_sources:
  - conda-forge/label/rapidjson_dev,conda-forge
``` 

> 1.1.0.postYYYYMMDD

I would generally be OK with this on the `main` label given that this is a build-time only dependency.

**Comment by zklaus** [comment]
I have updated the versioning scheme to the `1.1.0.postYYYYMMDD` variant, so if the CI goes green this could be merged.

As @xhochy explained, putting this on a `dev` label might be viable in this case because rapidjson is header only. That would be an opt-in variant, where packages need to actively decide to use a newer, date based version, whereas the current approach is opt-out, where packages will automatically receive the newer, date based versions unless they put an upper pin.

@timkpaine, is it ok as is, or would you prefer the opt-in variant of a dev label?

**Comment by timkpaine** [comment]
I like this as-is

**Review by xhochy — APPROVED** [review]

**Comment by zklaus** [comment]
Thanks @xhochy, @timkpaine :tada:

**Pull request merged**